### ISSUE TITLE
pragma no cover fails with elif

### ISSUE THREAD

**Issue by artgoldberg** [coveragepy/coveragepy#697]
Hi Ned, et al., 
Following https://coverage.readthedocs.io/en/coverage-4.3.4/excluding.html#branch-coverage, I expect 
```
    def test(self):
        if False:
            pass    # 1
        elif True:  # 2
            pass
        else:    # pragma no cover
            pass    # 3
        pass # 4
```
to not report that branch at # 2 fails to branch to the else. This works in the above example, but in a piece of production code it reports that the branch at # 2 fails to go to #4 instead:
![failed no cover](https://user-images.githubusercontent.com/33882/44727025-0bdb9380-aaa7-11e8-976b-d627e2b3d7d3.png)

unfortunately, I don't know how to reproduce this in a minimal snippet.

thanks
Arthur

**Comment by nedbat** [comment]
Arthur, thanks.  This seems to reproduce it:
```
$ cat bug697.py
def run():
    a = 0
    for i in range(10):
        if 0 <= i <= 3:
            a += 1
        elif i > 3:
            a += 10
        else:   # pragma: no cover
            pass
        a += 0
    return a

print(run())

$ coverage run --branch bug697.py
64

$ coverage report -m
Name        Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------
bug697.py      10      0      6      1    94%   6->10
```
The problem seems to be that the else:pass is optimized away, so the actual jumping seen by Python is to a line that is not excluded.  If we change the pass, then the problem goes away:
```
$ cat bug697_fixed.py
def run():
    a = 0
    for i in range(10):
        if 0 <= i <= 3:
            a += 1
        elif i > 3:
            a += 10
        else:   # pragma: no cover
            a += 1000
        a += 0
    return a

print(run())

$ coverage run --branch bug697_fixed.py
64

$ coverage report -m
Name              Stmts   Miss Branch BrPart  Cover   Missing
-------------------------------------------------------------
bug697_fixed.py      10      0      4      0   100%
```
This is related to a number of other issues reported against coverage that have to do with the peephole optimizer changing jumps to jumps.  Until Python gets a way to fully disable the optimizer, these problems will persist.  Commenting on this CPython ticket might push it forward: http://bugs.python.org/issue2506 "Add mechanism to disable optimizations".

**Comment by nedbat** [comment]
This is now fixed by Python 3.10

**Comment by artgoldberg** [comment]
Cool.

From: Ned Batchelder ***@***.***>
Reply-To: nedbat/coveragepy ***@***.***>
Date: Tuesday, November 2, 2021 at 6:43 AM
To: nedbat/coveragepy ***@***.***>
Cc: "Goldberg, Arthur P" ***@***.***>, Author ***@***.***>
Subject: Re: [nedbat/coveragepy] pragma no cover fails with elif (#697)

USE CAUTION: External Message.

This is now fixed by Python 3.10

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_nedbat_coveragepy_issues_697-23issuecomment-2D957321344&d=DwMCaQ&c=shNJtf5dKgNcPZ6Yh64b-A&r=p26xYuDOPpwD_1KMz0HcEpDitWzbwzrc9iNKx0lkRN8&m=-9q2ZP7H5kOIag5U3rucrV7yHihRbSk8MDWHy1u1KQ4&s=yF-hhUnUeNEwggf42H6YqcEnwnjSF5hSe9bppRW6YFc&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAAIIWRIZPXNF2TAX7T76IDUJ7FD5ANCNFSM4FR6TB2A&d=DwMCaQ&c=shNJtf5dKgNcPZ6Yh64b-A&r=p26xYuDOPpwD_1KMz0HcEpDitWzbwzrc9iNKx0lkRN8&m=-9q2ZP7H5kOIag5U3rucrV7yHihRbSk8MDWHy1u1KQ4&s=62y5KwbIHjL2j9Duo1rdSyI90G_QOWfv-cNWfSiudbc&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=shNJtf5dKgNcPZ6Yh64b-A&r=p26xYuDOPpwD_1KMz0HcEpDitWzbwzrc9iNKx0lkRN8&m=-9q2ZP7H5kOIag5U3rucrV7yHihRbSk8MDWHy1u1KQ4&s=5a7wQ-n5bWOwp-3lw5U081qeFUbdH8HSAxgla9UkYQc&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=shNJtf5dKgNcPZ6Yh64b-A&r=p26xYuDOPpwD_1KMz0HcEpDitWzbwzrc9iNKx0lkRN8&m=-9q2ZP7H5kOIag5U3rucrV7yHihRbSk8MDWHy1u1KQ4&s=sDeDVS0izDQw4J02ZnEmeKLYPTbMD-HN-Y1trhfEHq0&e=>.
